**AI ASSIGNMENT-03 20BCE2702**


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**Unzip the folder**



In [ ]:
!unzip '/content/drive/MyDrive/ssd/archive (1).zip'

Archive:  /content/drive/MyDrive/ssd/archive (1).zip
  inflating: test_data/test_data/blasti/DSC_6396.jpg  
  inflating: test_data/test_data/blasti/DSC_6397.jpg  
  inflating: test_data/test_data/blasti/DSC_6398.jpg  
  inflating: test_data/test_data/blasti/DSC_6399.jpg  
  inflating: test_data/test_data/blasti/DSC_6400.jpg  
  inflating: test_data/test_data/blasti/DSC_6401.jpg  
  inflating: test_data/test_data/blasti/DSC_6402.jpg  
  inflating: test_data/test_data/blasti/DSC_6403.jpg  
  inflating: test_data/test_data/blasti/DSC_6405.jpg  
  inflating: test_data/test_data/blasti/DSC_6406.jpg  
  inflating: test_data/test_data/blasti/DSC_6407.jpg  
  inflating: test_data/test_data/blasti/DSC_6408.jpg  
  inflating: test_data/test_data/blasti/DSC_6409.jpg  
  inflating: test_data/test_data/blasti/DSC_6410.jpg  
  inflating: test_data/test_data/blasti/DSC_6411.jpg  
  inflating: test_data/test_data/bonegl/DSC_4587.jpg  
  inflating: test_data/test_data/bonegl/DSC_4588.jpg  
  inflating:

In [ ]:
# Import req lib
from tensorflow.keras.layers import Dense,Flatten,Input
from tensorflow.keras.models import Model
from tensorflow.keras.preprocessing import image
from tensorflow.keras.preprocessing.image import ImageDataGenerator, load_img
import numpy as np

In [ ]:
train_path = "/content/train_data/train_data"
test_path = "/content/test_data/test_data"

**Data augumentation --> Converting one image into multiple images**

In [ ]:
train_gen = ImageDataGenerator(rescale=(1./255),horizontal_flip=True,shear_range=0.2)
test_gen = ImageDataGenerator(rescale=(1./255))  #--> (0 to 255) convert to (0 to 1)

In [ ]:
train = train_gen.flow_from_directory('/content/train_data/train_data',
                                      target_size=(224, 224),
                                      class_mode='categorical',
                                      batch_size=22)
test = test_gen.flow_from_directory('/content/test_data/test_data',
                                    target_size=(224, 224),
                                      class_mode='categorical',
                                      batch_size=22)

Found 150 images belonging to 16 classes.
Found 157 images belonging to 16 classes.


In [ ]:
train.class_indices

{'blasti': 0,
 'bonegl': 1,
 'brhkyt': 2,
 'cbrtsh': 3,
 'cmnmyn': 4,
 'gretit': 5,
 'hilpig': 6,
 'himbul': 7,
 'himgri': 8,
 'hsparo': 9,
 'indvul': 10,
 'jglowl': 11,
 'lbicrw': 12,
 'mgprob': 13,
 'rebimg': 14,
 'wcrsrt': 15}

**Transfer Learning**

In [ ]:
"""
1. VGG16
2. ResNet50
3. Inseption
4. Xception
"""

'\n1. VGG16\n2. ResNet50\n3. Inseption\n4. Xception\n'

**VGG16**

In [ ]:
from tensorflow.keras.applications.vgg16 import VGG16, preprocess_input

In [ ]:
# Adding the preprocessing layer to the front of vgg

vgg = VGG16(include_top=False,weights='imagenet',input_shape=(224,224,3))

58889256/58889256 [==============================] - 4s 0us/step


In [ ]:
# Train model with existing weights

for layer in vgg.layers:
  print(layer)

In [ ]:

# Train model with existing weights

for layer in vgg.layers:
  layer.trainable=False

In [ ]:
x = Flatten()(vgg.output)

In [ ]:
# output layer

prediction = Dense(16,activation='softmax')(x)

In [ ]:
# Create Vgg16 model

model = Model(inputs=vgg.input,outputs=prediction)

In [ ]:
model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 224, 224, 3)]     0         
                                                                 
 block1_conv1 (Conv2D)       (None, 224, 224, 64)      1792      
                                                                 
 block1_conv2 (Conv2D)       (None, 224, 224, 64)      36928     
                                                                 
 block1_pool (MaxPooling2D)  (None, 112, 112, 64)      0         
                                                                 
 block2_conv1 (Conv2D)       (None, 112, 112, 128)     73856     
                                                                 
 block2_conv2 (Conv2D)       (None, 112, 112, 128)     147584    
                                                                 
 block2_pool (MaxPooling2D)  (None, 56, 56, 128)       0     

In [ ]:
model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])

In [ ]:
model.fit_generator(train,validation_data=test,epochs=10,steps_per_epoch=len(train),
                    validation_steps=len(test))

<ipython-input-17-240d737b1fe6>:1: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.
  model.fit_generator(train,validation_data=test,epochs=10,steps_per_epoch=len(train),


Epoch 1/10
7/7 [==============================] - 106s 15s/step - loss: 3.0539 - accuracy: 0.2667 - val_loss: 2.8776 - val_accuracy: 0.2484
Epoch 2/10
7/7 [==============================] - 121s 19s/step - loss: 1.3714 - accuracy: 0.5800 - val_loss: 2.7200 - val_accuracy: 0.3248
Epoch 3/10
7/7 [==============================] - 120s 19s/step - loss: 0.6742 - accuracy: 0.8133 - val_loss: 2.6600 - val_accuracy: 0.3121
Epoch 4/10
7/7 [==============================] - 120s 19s/step - loss: 0.3091 - accuracy: 0.9333 - val_loss: 2.9872 - val_accuracy: 0.3503
Epoch 5/10
7/7 [==============================] - 88s 14s/step - loss: 0.1624 - accuracy: 0.9667 - val_loss: 3.0162 - val_accuracy: 0.3694
Epoch 6/10
7/7 [==============================] - 90s 14s/step - loss: 0.1165 - accuracy: 0.9733 - val_loss: 2.9278 - val_accuracy: 0.3694
Epoch 7/10
7/7 [==============================] - 88s 14s/step - loss: 0.0873 - accuracy: 0.9867 - val_loss: 2.9507 - val_accuracy: 0.3567
Epoch 8/10
7/7 [=======

**This vgg16 model gives the accuracy rate of 42 so we can try with other transfer learning methods like Resnet50,Inception and Xception.**

**ResNet50**

In [ ]:
from tensorflow.keras.applications.resnet50 import ResNet50


In [ ]:
resnet = ResNet50(include_top=False,input_shape=(224,224,3))


In [ ]:
for layer in resnet.layers:
  layer.trainable=False

In [ ]:
for layer in resnet.layers:
  print(layer)

In [ ]:
x = Flatten()(resnet.output)


In [ ]:
out = Dense(16, activation='softmax')(x)


In [ ]:
res_model = Model(inputs=resnet.input,outputs=out)


In [ ]:
res_model.summary()


Model: "model_3"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_4 (InputLayer)           [(None, 224, 224, 3  0           []                               
                                )]                                                                
                                                                                                  
 conv1_pad (ZeroPadding2D)      (None, 230, 230, 3)  0           ['input_4[0][0]']                
                                                                                                  
 conv1_conv (Conv2D)            (None, 112, 112, 64  9472        ['conv1_pad[0][0]']              
                                )                                                                 
                                                                                            

In [ ]:
res_model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])


In [ ]:
res_model.fit_generator(train,epochs=10,validation_data=test,steps_per_epoch=len(train),
              validation_steps=len(test))

<ipython-input-49-ff17a629f64b>:1: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.
  res_model.fit_generator(train,epochs=10,validation_data=test,steps_per_epoch=len(train),


Epoch 1/10
7/7 [==============================] - 95s 14s/step - loss: 13.1771 - accuracy: 0.0533 - val_loss: 15.4456 - val_accuracy: 0.1210
Epoch 2/10
7/7 [==============================] - 94s 14s/step - loss: 11.4387 - accuracy: 0.1333 - val_loss: 9.5339 - val_accuracy: 0.1146
Epoch 3/10
7/7 [==============================] - 93s 14s/step - loss: 8.7051 - accuracy: 0.1600 - val_loss: 7.6318 - val_accuracy: 0.1019
Epoch 4/10
7/7 [==============================] - 84s 13s/step - loss: 6.2572 - accuracy: 0.1800 - val_loss: 8.1319 - val_accuracy: 0.1465
Epoch 5/10
7/7 [==============================] - 84s 13s/step - loss: 4.9322 - accuracy: 0.3133 - val_loss: 5.9984 - val_accuracy: 0.1592
Epoch 6/10
7/7 [==============================] - 85s 13s/step - loss: 3.6370 - accuracy: 0.2200 - val_loss: 4.7013 - val_accuracy: 0.2229
Epoch 7/10
7/7 [==============================] - 119s 19s/step - loss: 2.4372 - accuracy: 0.3933 - val_loss: 3.8848 - val_accuracy: 0.2229
Epoch 8/10
7/7 [=======

**Resnet also gives very less accuracy of 21 , so we can go with other methods like Inception and Xception.**

**Inseption**

In [ ]:
from tensorflow.keras.applications.inception_v3 import InceptionV3


In [ ]:
inception = InceptionV3(include_top=False, input_shape=(224, 224, 3))


87910968/87910968 [==============================] - 0s 0us/step


In [ ]:

for layer in inception.layers:
    layer.trainable = False



In [ ]:

for layer in inception.layers:
    print(layer)



In [ ]:
x = Flatten()(inception.output)


In [ ]:
out = Dense(16, activation='softmax')(x)



In [ ]:
inception_model = Model(inputs=inception.input, outputs=out)


In [ ]:
inception_model.summary()



Model: "model_4"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_5 (InputLayer)           [(None, 224, 224, 3  0           []                               
                                )]                                                                
                                                                                                  
 conv2d (Conv2D)                (None, 111, 111, 32  864         ['input_5[0][0]']                
                                )                                                                 
                                                                                                  
 batch_normalization (BatchNorm  (None, 111, 111, 32  96         ['conv2d[0][0]']                 
 alization)                     )                                                           

In [ ]:
inception_model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])



In [ ]:
inception_model.fit_generator(train, epochs=10, validation_data=test, steps_per_epoch=len(train), validation_steps=len(test))


<ipython-input-61-44ad6cecd50c>:1: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.
  inception_model.fit_generator(train, epochs=10, validation_data=test, steps_per_epoch=len(train), validation_steps=len(test))


Epoch 1/10
7/7 [==============================] - 114s 17s/step - loss: 16.6311 - accuracy: 0.1733 - val_loss: 15.5126 - val_accuracy: 0.2357
Epoch 2/10
7/7 [==============================] - 84s 13s/step - loss: 8.2248 - accuracy: 0.5533 - val_loss: 17.1827 - val_accuracy: 0.3185
Epoch 3/10
7/7 [==============================] - 85s 13s/step - loss: 2.0008 - accuracy: 0.8200 - val_loss: 17.4271 - val_accuracy: 0.2420
Epoch 4/10
7/7 [==============================] - 84s 13s/step - loss: 1.0563 - accuracy: 0.8733 - val_loss: 12.1821 - val_accuracy: 0.4459
Epoch 5/10
7/7 [==============================] - 119s 19s/step - loss: 0.5725 - accuracy: 0.9400 - val_loss: 11.6656 - val_accuracy: 0.4650
Epoch 6/10
7/7 [==============================] - 83s 13s/step - loss: 0.1979 - accuracy: 0.9800 - val_loss: 13.8035 - val_accuracy: 0.3885
Epoch 7/10
7/7 [==============================] - 83s 13s/step - loss: 0.1495 - accuracy: 0.9800 - val_loss: 12.3544 - val_accuracy: 0.4395
Epoch 8/10
7/7 [=

In [ ]:
# Testing --> Since we are continously getting the accuracy of 40 in vgg16 as well as here in this model so we randomly test the model of cnn.

import numpy as np
from tensorflow.keras.preprocessing import image

In [ ]:
# Testing 1
img1 = image.load_img('/content/test_data/test_data/cbrtsh/_D32_10310.jpg',target_size=(224,224))
img1 = image.img_to_array(img1)
img1 = np.expand_dims(img1,axis=0)
pred = np.argmax(model.predict(img1))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 1s 956ms/step
7
himbul


In [ ]:
# Testing 2
img2 = image.load_img('/content/test_data/test_data/brhkyt/D72_0473.jpg',target_size=(224,224))
img2 = image.img_to_array(img2)
img2 = np.expand_dims(img2,axis=0)
pred = np.argmax(model.predict(img2))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']

print(output[pred])

1/1 [==============================] - 0s 19ms/step
2
brhkyt


In [ ]:
# Testing 3
img3 = image.load_img('/content/test_data/test_data/bonegl/DSC_4587.jpg',target_size=(224,224))
img3 = image.img_to_array(img3)
img3 = np.expand_dims(img3,axis=0)
pred = np.argmax(model.predict(img3))
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']

print(pred)

print(output[pred])

1/1 [==============================] - 0s 21ms/step
7
himbul


In [ ]:
# Testing 4
img4 = image.load_img('/content/test_data/test_data/blasti/DSC_6396.jpg',target_size=(224,224))
img4 = image.img_to_array(img4)
img4 = np.expand_dims(img4,axis=0)
pred = np.argmax(model.predict(img4))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 22ms/step
0
blasti


In [ ]:
# Testing 5
img5 = image.load_img('/content/test_data/test_data/rebimg/100_5756.JPG',target_size=(224,224))
img5 = image.img_to_array(img5)
img5 = np.expand_dims(img5,axis=0)
pred = np.argmax(model.predict(img5))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 27ms/step
14
wcrsrt


In [ ]:
img6 = image.load_img('/content/test_data/test_data/jglowl/_D32_10587.jpg',target_size=(224,224))
img6 = image.img_to_array(img6)
img6 = np.expand_dims(img6,axis=0)
pred = np.argmax(model.predict(img6))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 20ms/step
14
wcrsrt


In [ ]:
img7 = image.load_img('/content/test_data/test_data/cmnmyn/DSC_2443.jpg',target_size=(224,224))
img7 = image.img_to_array(img7)
img7 = np.expand_dims(img7,axis=0)
pred = np.argmax(model.predict(img7))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 19ms/step
7
himbul


In [ ]:
img8 = image.load_img('/content/test_data/test_data/gretit/D72_0693.jpg',target_size=(224,224))
img8 = image.img_to_array(img8)
img8 = np.expand_dims(img8,axis=0)
pred = np.argmax(model.predict(img8))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 21ms/step
4
cmnmyn


In [ ]:
img9 = image.load_img('/content/test_data/test_data/hilpig/DSC_6359.jpg',target_size=(224,224))
img9 = image.img_to_array(img9)
img9 = np.expand_dims(img9,axis=0)
pred = np.argmax(model.predict(img9))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 28ms/step
7
himbul


In [ ]:
img10 = image.load_img('/content/test_data/test_data/himbul/_D32_10242.jpg',target_size=(224,224))
img10 = image.img_to_array(img10)
img10 = np.expand_dims(img10,axis=0)
pred = np.argmax(model.predict(img10))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 71ms/step
4
cmnmyn


In [ ]:
img11 = image.load_img('/content/test_data/test_data/himgri/IMG_5384.JPG',target_size=(224,224))
img11 = image.img_to_array(img11)
img11 = np.expand_dims(img11,axis=0)
pred = np.argmax(model.predict(img11))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 19ms/step
7
himbul


In [ ]:
img12 = image.load_img('/content/test_data/test_data/hsparo/DSC_4521.jpg',target_size=(224,224))
img12 = image.img_to_array(img12)
img12 = np.expand_dims(img12,axis=0)
pred = np.argmax(model.predict(img12))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 21ms/step
14
wcrsrt


In [ ]:
img13 = image.load_img('/content/test_data/test_data/indvul/DSC_1263.jpg',target_size=(224,224))
img13 = image.img_to_array(img13)
img13 = np.expand_dims(img13,axis=0)
pred = np.argmax(model.predict(img13))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 27ms/step
14
wcrsrt


In [ ]:
img14 = image.load_img('/content/test_data/test_data/lbicrw/100_4917.JPG',target_size=(224,224))
img14 = image.img_to_array(img14)
img14 = np.expand_dims(img14,axis=0)
pred = np.argmax(model.predict(img14))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 18ms/step
12
mgprob


**Xception**

In [ ]:
from tensorflow.keras.applications.xception import Xception



In [ ]:
xception = Xception(include_top=False, input_shape=(224, 224, 3))



83683744/83683744 [==============================] - 0s 0us/step


In [ ]:
for layer in xception.layers:
    layer.trainable = False



In [ ]:


for layer in xception.layers:
    print(layer)


In [ ]:
x = Flatten()(xception.output)

In [ ]:
out = Dense(16, activation='softmax')(x)


In [ ]:
xception_model = Model(inputs=xception.input, outputs=out)

In [ ]:
xception_model.summary()


Model: "model_5"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_6 (InputLayer)           [(None, 224, 224, 3  0           []                               
                                )]                                                                
                                                                                                  
 block1_conv1 (Conv2D)          (None, 111, 111, 32  864         ['input_6[0][0]']                
                                )                                                                 
                                                                                                  
 block1_conv1_bn (BatchNormaliz  (None, 111, 111, 32  128        ['block1_conv1[0][0]']           
 ation)                         )                                                           

In [ ]:
xception_model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])


In [ ]:
xception_model.fit_generator(train, epochs=10, validation_data=test, steps_per_epoch=len(train), validation_steps=len(test))


<ipython-input-88-66ed3aaf8f62>:1: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.
  xception_model.fit_generator(train, epochs=10, validation_data=test, steps_per_epoch=len(train), validation_steps=len(test))


Epoch 1/10
7/7 [==============================] - 108s 16s/step - loss: 7.3407 - accuracy: 0.3533 - val_loss: 8.7848 - val_accuracy: 0.4076
Epoch 2/10
7/7 [==============================] - 86s 13s/step - loss: 1.6053 - accuracy: 0.7800 - val_loss: 9.6903 - val_accuracy: 0.4395
Epoch 3/10
7/7 [==============================] - 87s 14s/step - loss: 0.5143 - accuracy: 0.9267 - val_loss: 11.0276 - val_accuracy: 0.4459
Epoch 4/10
7/7 [==============================] - 119s 19s/step - loss: 0.3601 - accuracy: 0.9600 - val_loss: 11.7372 - val_accuracy: 0.4395
Epoch 5/10
7/7 [==============================] - 119s 19s/step - loss: 0.3430 - accuracy: 0.9667 - val_loss: 12.2848 - val_accuracy: 0.4395
Epoch 6/10
7/7 [==============================] - 84s 13s/step - loss: 0.1530 - accuracy: 0.9533 - val_loss: 13.6495 - val_accuracy: 0.4586
Epoch 7/10
7/7 [==============================] - 119s 19s/step - loss: 0.1361 - accuracy: 0.9800 - val_loss: 15.3631 - val_accuracy: 0.4841
Epoch 8/10
7/7 [==

In [ ]:
# Testing 1 -->This xception model also has the accuracy of 40 so we test the model using examples taken from the test data which exactly gives the same output as inception model.
img1 = image.load_img('/content/test_data/test_data/cbrtsh/_D32_10310.jpg',target_size=(224,224))
img1 = image.img_to_array(img1)
img1 = np.expand_dims(img1,axis=0)
pred = np.argmax(model.predict(img1))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 21ms/step
7
himbul


In [ ]:
# Testing 2
img2 = image.load_img('/content/test_data/test_data/brhkyt/D72_0473.jpg',target_size=(224,224))
img2 = image.img_to_array(img2)
img2 = np.expand_dims(img2,axis=0)
pred = np.argmax(model.predict(img2))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']

print(output[pred])

1/1 [==============================] - 0s 30ms/step
2
brhkyt


In [ ]:
# Testing 3
img3 = image.load_img('/content/test_data/test_data/bonegl/DSC_4587.jpg',target_size=(224,224))
img3 = image.img_to_array(img3)
img3 = np.expand_dims(img3,axis=0)
pred = np.argmax(model.predict(img3))
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']

print(pred)

print(output[pred])

1/1 [==============================] - 0s 19ms/step
7
himbul


In [ ]:
# Testing 4
img4 = image.load_img('/content/test_data/test_data/blasti/DSC_6396.jpg',target_size=(224,224))
img4 = image.img_to_array(img4)
img4 = np.expand_dims(img4,axis=0)
pred = np.argmax(model.predict(img4))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 22ms/step
0
blasti


In [ ]:
# Testing 5
img5 = image.load_img('/content/test_data/test_data/rebimg/100_5756.JPG',target_size=(224,224))
img5 = image.img_to_array(img5)
img5 = np.expand_dims(img5,axis=0)
pred = np.argmax(model.predict(img5))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 20ms/step
14
wcrsrt


In [ ]:
img6 = image.load_img('/content/test_data/test_data/jglowl/_D32_10587.jpg',target_size=(224,224))
img6 = image.img_to_array(img6)
img6 = np.expand_dims(img6,axis=0)
pred = np.argmax(model.predict(img6))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 30ms/step
14
wcrsrt


In [ ]:
img7 = image.load_img('/content/test_data/test_data/cmnmyn/DSC_2443.jpg',target_size=(224,224))
img7 = image.img_to_array(img7)
img7 = np.expand_dims(img7,axis=0)
pred = np.argmax(model.predict(img7))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 19ms/step
7
himbul


In [ ]:
img8 = image.load_img('/content/test_data/test_data/gretit/D72_0693.jpg',target_size=(224,224))
img8 = image.img_to_array(img8)
img8 = np.expand_dims(img8,axis=0)
pred = np.argmax(model.predict(img8))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 20ms/step
4
cmnmyn


In [ ]:
img9 = image.load_img('/content/test_data/test_data/hilpig/DSC_6359.jpg',target_size=(224,224))
img9 = image.img_to_array(img9)
img9 = np.expand_dims(img9,axis=0)
pred = np.argmax(model.predict(img9))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 24ms/step
7
himbul


In [ ]:
img10 = image.load_img('/content/test_data/test_data/himbul/_D32_10242.jpg',target_size=(224,224))
img10 = image.img_to_array(img10)
img10 = np.expand_dims(img10,axis=0)
pred = np.argmax(model.predict(img10))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 19ms/step
4
cmnmyn


In [ ]:
img11 = image.load_img('/content/test_data/test_data/himgri/IMG_5384.JPG',target_size=(224,224))
img11 = image.img_to_array(img11)
img11 = np.expand_dims(img11,axis=0)
pred = np.argmax(model.predict(img11))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 22ms/step
7
himbul


In [ ]:
img12 = image.load_img('/content/test_data/test_data/hsparo/DSC_4521.jpg',target_size=(224,224))
img12 = image.img_to_array(img12)
img12 = np.expand_dims(img12,axis=0)
pred = np.argmax(model.predict(img12))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 29ms/step
14
wcrsrt


In [ ]:
img13 = image.load_img('/content/test_data/test_data/indvul/DSC_1263.jpg',target_size=(224,224))
img13 = image.img_to_array(img13)
img13 = np.expand_dims(img13,axis=0)
pred = np.argmax(model.predict(img13))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

1/1 [==============================] - 0s 27ms/step
14
wcrsrt


In [ ]:
img14 = image.load_img('/content/test_data/test_data/lbicrw/100_4917.JPG',target_size=(224,224))
img14 = image.img_to_array(img14)
img14 = np.expand_dims(img14,axis=0)
print(img14)
pred = np.argmax(model.predict(img14))
print(pred)
output = ['blasti','bonegl','brhkyt','cbrtsh','cmnmyn','gretit','hilpig','himbul','hsparo','indvul','jglowl','lbicrw','mgprob','rebimg','wcrsrt']
print(output[pred])

[[[[115. 120. 126.]
   [116. 120. 129.]
   [114. 118. 127.]
   ...
   [115. 119. 128.]
   [115. 117. 129.]
   [116. 119. 128.]]

  [[115. 120. 126.]
   [116. 120. 129.]
   [114. 118. 127.]
   ...
   [117. 121. 130.]
   [115. 117. 129.]
   [116. 119. 128.]]

  [[118. 123. 129.]
   [117. 120. 129.]
   [113. 122. 129.]
   ...
   [115. 119. 128.]
   [115. 117. 129.]
   [115. 117. 129.]]

  ...

  [[118. 117. 122.]
   [117. 116. 121.]
   [118. 119. 123.]
   ...
   [114. 113. 121.]
   [115. 116. 121.]
   [113. 113. 121.]]

  [[120. 121. 125.]
   [117. 118. 123.]
   [119. 118. 124.]
   ...
   [115. 114. 122.]
   [116. 115. 123.]
   [113. 114. 119.]]

  [[116. 117. 121.]
   [119. 120. 125.]
   [118. 117. 123.]
   ...
   [115. 114. 122.]
   [116. 115. 123.]
   [114. 115. 120.]]]]
1/1 [==============================] - 0s 23ms/step
12
mgprob


**Result:**
**We get the accuracy of 40 percent in vgg16,Inception and Xception also if we test the model twice with the same examples gives the same output.
This accuracy is because we have 150 images in training dataset and 157 images in testing dataset which causes the model to overfit.
Overfitting happens due to several reasons, such as: • The training data size is too small and does not contain enough data samples to accurately represent all possible input data values.
In the model testing we get two to three images same to the class defined in the dataset out of 16 classess.**